### ISSUE TITLE
merge seems broken on encrypted files?

### ISSUE THREAD

**Issue by indirect** [twpayne/chezmoi#2175]
## Describe the bug

I noticed recently (perhaps ever since textconv support landed?) that `chezmoi merge` fails on files that are encrypted, with an error like "no such file or directory".

## To reproduce

1. Compile the latest `chezmoi` from 71a65f4da3ce45ff6b6b2e1aa4c14d3aa62a4efa.
2. Set up an empty chezmoi-managed directory with encryption:

```bash
$ age-keygen -o chezmoi_key.txt
Public key: age1r33xjzny7ev9th5wxg5ugx6d0mu5laq9suhfc5k24nd65wqxn4nq9s9l8u
$ echo 'encryption = "age"
[age]
    identity = "chezmoi_key.txt"
    recipient = "age1r33xjzny7ev9th5wxg5ugx6d0mu5laq9suhfc5k24nd65wqxn4nq9s9l8u"' > chezmoi.toml
$ chezmoi --cache cache --config chezmoi.toml -W . -S source init
```

4.  Try to merge an encrypted file

```bash
$ echo "some content" > example.txt
$ chezmoi --cache cache --config chezmoi.toml -W . -S source add --encrypt example.txt
$ chezmoi --cache cache --config chezmoi.toml -W . -S source merge example.txt
chezmoi: stat /var/folders/lh/b8tnjh816b9bzfskhfbq16v00000gn/T/chezmoi-merge-plaintext3053760180/[...]: no such file or directory
```

## Expected behavior

I expected my editor to open with a 3-way merge, like when I do this with a non-encrypted file.

## Output of command with the `--debug` flag

<details>

```console
❯ chezmoi --cache cache --config chezmoi.toml -W . -S source merge example.txt --debug
2022-06-22T23:19:02-07:00 INF persistentPreRunRootE args=["chezmoi","--cache","cache","--config","chezmoi.toml","-W",".","-S","source","merge","example.txt","--debug"] goVersion=go1.18.3 version={"builtBy":"","commit":"","date":"","version":""}
2022-06-22T23:19:02-07:00 INF Stat component=system name=/Users/andre/src/indirect/testcases/age-merge/source
2022-06-22T23:19:02-07:00 ERR ReadFile error="open /Users/andre/src/indirect/testcases/age-merge/source/.chezmoiversion: no such file or directory" component=system data= name=/Users/andre/src/indirect/testcases/age-merge/source/.chezmoiversion size=0
2022-06-22T23:19:02-07:00 ERR ReadFile error="open /Users/andre/src/indirect/testcases/age-merge/source/.chezmoiroot: no such file or directory" component=system data= name=/Users/andre/src/indirect/testcases/age-merge/source/.chezmoiroot size=0
2022-06-22T23:19:02-07:00 INF Stat component=system name=/Users/andre/src/indirect/testcases/age-merge/source
2022-06-22T23:19:02-07:00 INF Stat component=system name=/Users/andre/src/indirect/testcases/age-merge/source
2022-06-22T23:19:02-07:00 INF ReadDir component=system name=/Users/andre/src/indirect/testcases/age-merge/source
2022-06-22T23:19:02-07:00 INF ReadDir component=system name=/Users/andre/src/indirect/testcases/age-merge/source/src
2022-06-22T23:19:02-07:00 INF ReadDir component=system name=/Users/andre/src/indirect/testcases/age-merge/source/src/indirect
2022-06-22T23:19:02-07:00 INF ReadDir component=system name=/Users/andre/src/indirect/testcases/age-merge/source/src/indirect/testcases
2022-06-22T23:19:02-07:00 INF ReadDir component=system name=/Users/andre/src/indirect/testcases/age-merge/source/src/indirect/testcases/age-merge
2022-06-22T23:19:02-07:00 INF MkdirTemp tempDir=/var/folders/lh/b8tnjh816b9bzfskhfbq16v00000gn/T/chezmoi-merge-plaintext1023722716
2022-06-22T23:19:02-07:00 INF ReadFile component=system data="-----BEGIN AGE ENCRYPTED FILE-----\nYWdlLWVuY3J5cHRpb24ub3JnL3YxC..." name=/Users/andre/src/indirect/testcases/age-merge/source/src/indirect/testcases/age-merge/encrypted_example.txt.age size=357
2022-06-22T23:19:02-07:00 INF Output args=["age","--decrypt","--identity","/Users/andre/src/indirect/testcases/age-merge/chezmoi_key.txt"] duration=6.640125ms output="some content\n" path=/opt/homebrew/bin/age size=13
2022-06-22T23:19:02-07:00 INF Decrypt ciphertext="-----BEGIN AGE ENCRYPTED FILE-----\nYWdlLWVuY3J5cHRpb24ub3JnL3YxC..." component=encryption plaintext="some content\n"
2022-06-22T23:19:02-07:00 ERR WriteFile error="stat /var/folders/lh/b8tnjh816b9bzfskhfbq16v00000gn/T/chezmoi-merge-plaintext1023722716/src/indirect/testcases/age-merge: no such file or directory" component=system data="some content\n" name=/var/folders/lh/b8tnjh816b9bzfskhfbq16v00000gn/T/chezmoi-merge-plaintext1023722716/src/indirect/testcases/age-merge/encrypted_example.txt.age perm=384 size=13
2022-06-22T23:19:02-07:00 INF RemoveAll tempDir=/var/folders/lh/b8tnjh816b9bzfskhfbq16v00000gn/T/chezmoi-merge-plaintext1023722716
chezmoi: stat /var/folders/lh/b8tnjh816b9bzfskhfbq16v00000gn/T/chezmoi-merge-plaintext1023722716/src/indirect/testcases/age-merge: no such file or directory
```

</details>

## Output of `chezmoi doctor`

<details>

```console
❯ chezmoi --cache cache --config chezmoi.toml -W . -S source doctor
RESULT    CHECK                MESSAGE
warning   version              dev
warning   latest-version       v2.18.0
ok        os-arch              darwin/arm64
ok        uname                Darwin gnk.local 21.6.0 Darwin Kernel Version 21.6.0: Mon Jun  6 19:48:18 PDT 2022; root:xnu-8020.140.36.131.1~1/RELEASE_ARM64_T6000 arm64
ok        go-version           go1.18.3 (gc)
ok        executable           ~/src/twpayne/chezmoi/chezmoi
ok        upgrade-method       replace-executable
failed    config-file          found ~/.config/chezmoi/chezmoi.toml, expected ~/src/indirect/testcases/age-merge/chezmoi.toml
ok        source-dir           ~/src/indirect/testcases/age-merge/source is a directory
ok        suspicious-entries   no suspicious entries
warning   working-tree         ~/src/indirect/testcases/age-merge is a git working tree (dirty)
ok        dest-dir             ~ is a directory
ok        shell-command        found /bin/zsh
ok        shell-args           /bin/zsh
ok        cd-command           found /bin/zsh
ok        cd-args              /bin/zsh
ok        edit-command         found /opt/homebrew/bin/mvim
ok        edit-args            mvim -f --nomru -c 'au VimLeave * !open -a Terminal'
info      diff-command         not set
ok        umask                022
ok        git-command          found /opt/homebrew/bin/git, version 2.36.1
ok        merge-command        found /opt/homebrew/bin/vimdiff
ok        age-command          found /opt/homebrew/bin/age, version 1.0.0
ok        gpg-command          found /opt/homebrew/bin/gpg, version 2.3.6
info      pinentry-command     not set
ok        1password-command    found /usr/local/bin/op, version 2.4.1
info      bitwarden-command    bw not found in $PATH
info      gopass-command       gopass not found in $PATH
info      keepassxc-command    keepassxc-cli not found in $PATH
info      keeper-command       keeper not found in $PATH
info      keepassxc-db         not set
info      lastpass-command     lpass not found in $PATH
info      pass-command         pass not found in $PATH
info      vault-command        vault not found in $PATH
info      secret-command       not set
```

</details>

## Additional context

Based on the debug output, it seems like the tmp dir is getting cleaned up before the file is getting opened? Or maybe the file isn't getting written in the first place, because the tmp dir doesn't actually exist yet. Or maybe I'm doing something very wrong, somehow.

**Comment by twpayne** [comment]
Hmm, this is weird. There's an [explicit test for merging encrypted files](https://github.com/twpayne/chezmoi/blob/v2.18.0/pkg/cmd/testdata/scripts/mergeencrypted_unix.txt) which passes, but maybe the test is not correct. textconv does not touch the temporary directory (it feeds a file into a command's standard input and slurps the standard output) so that's unlikely to be the cause. This does look like a genuine bug in chezmoi, however. Investigating.

**Comment by twpayne** [comment]
I've tried to reproduce this but I have not been successful. I've tried:

* Following your steps to reproduce on my macOS machine.
* A test script in #2176.
* Running the following in a fresh empty Docker container (`docker run -it --rm ubuntu:latest /bin/bash`):

```sh
#!/bin/bash

cd
apt update
apt install -y age curl git vim
sh -c "$(curl -fsSL chezmoi.io/get)"
export PATH=$HOME/bin:$PATH
public_key=$(age-keygen -o chezmoi_key.txt 2>&1 | awk '{print $3}'
mkdir -p .config/chezmoi
echo 'encryption = "age"' >> .config/chezmoi/chezmoi.toml
echo '[age]' >> .config/chezmoi/chezmoi.toml
echo "  identity = \"$HOME/chezmoi_key.txt\"" >> .config/chezmoi/chezmoi.toml
echo "  recipient = \"$public_key\"" >> .config/chezmoi/chezmoi.toml
echo 'some content' > example.txt
chezmoi add --encrypt example.txt
chezmoi merge example.txt
```

Do you have a case that reproduces the bug on a new, empty machine?

**Comment by indirect** [comment]
Thank you for the quick response and the very thorough investigation—I especially appreciate you adding a testcase. I was able to use your script to verify that a different user on my machine doesn't have the problem. Time to narrow down what file or configuration is causing the issue.

**Comment by indirect** [comment]
AHA @segiddins spent some time debugging with me and we figured out that it breaks when the file is inside a directory. Here's a script that reproduces the bug inside an empty container (`docker run -it --rm ubuntu:latest /bin/bash`):

```bash
#!/bin/bash

cd
apt update
apt install -y age curl git vim
sh -c "$(curl -fsSL chezmoi.io/get)"
export PATH=$HOME/bin:$PATH
public_key=$(age-keygen -o chezmoi_key.txt 2>&1 | awk '{print $3}')
mkdir -p .config/chezmoi
echo 'encryption = "age"' >> .config/chezmoi/chezmoi.toml
echo '[age]' >> .config/chezmoi/chezmoi.toml
echo "  identity = \"$HOME/chezmoi_key.txt\"" >> .config/chezmoi/chezmoi.toml
echo "  recipient = \"$public_key\"" >> .config/chezmoi/chezmoi.toml
mkdir .directory
echo 'some content' > .directory/example.txt
chezmoi add --encrypt .directory/example.txt
chezmoi merge --verbose --debug .directory/example.txt
```

Here's the relevant output:

```
[...]
2022-06-29T03:48:51Z INF Stat component=system name=/root/.local/share/chezmoi
2022-06-29T03:48:51Z INF Stat component=system name=/root/.local/share/chezmoi
2022-06-29T03:48:51Z INF ReadDir component=system name=/root/.local/share/chezmoi
2022-06-29T03:48:51Z INF ReadDir component=system name=/root/.local/share/chezmoi/dot_directory
2022-06-29T03:48:51Z INF MkdirTemp tempDir=/tmp/chezmoi-merge-plaintext1661122327
2022-06-29T03:48:51Z INF ReadFile component=system data="-----BEGIN AGE ENCRYPTED FILE-----\nYWdlLWVuY3J5cHRpb24ub3JnL3YxC..." name=/root/.local/share/chezmoi/dot_directory/encrypted_example.txt.age size=357
2022-06-29T03:48:51Z INF Output args=["age","--decrypt","--identity","/root/chezmoi_key.txt"] duration=2.579208ms output="some content\n" path=/usr/bin/age size=13
2022-06-29T03:48:51Z INF Decrypt ciphertext="-----BEGIN AGE ENCRYPTED FILE-----\nYWdlLWVuY3J5cHRpb24ub3JnL3YxC..." component=encryption plaintext="some content\n"
2022-06-29T03:48:51Z ERR WriteFile error="stat /tmp/chezmoi-merge-plaintext1661122327/dot_directory: no such file or directory" component=system data="some content\n" name=/tmp/chezmoi-merge-plaintext1661122327/dot_directory/encrypted_example.txt.age perm=384 size=13
2022-06-29T03:48:51Z INF RemoveAll tempDir=/tmp/chezmoi-merge-plaintext1661122327
chezmoi: stat /tmp/chezmoi-merge-plaintext1661122327/dot_directory: no such file or directory
```

**Comment by twpayne** [comment]
Thanks, this indeed reproduces the bug. See #2183.

**Comment by twpayne** [comment]
Thank you so much @indirect and @segiddins for reporting this, your patience, and your thorough investigation. This should be fixed by #2183.
